waku/v2/protocol/peer_exchange/shard_lru.go

```diff
@@ -0,0 +1,167 @@
+package peer_exchange
+
+import (
+	"container/list"
+	"fmt"
+	"math/rand"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	wenr "github.com/waku-org/go-waku/waku/v2/protocol/enr"
+	"github.com/waku-org/go-waku/waku/v2/utils"
+)
+
+type ShardInfo struct {
+	cluster uint16
+	shard   uint16
+}
+type shardLRU struct {
+	size       int // number of nodes allowed per shard
+	idToNode   map[enode.ID][]*list.Element
+	shardNodes map[ShardInfo]*list.List
+	rng        *rand.Rand
+	mu         sync.RWMutex
+}
+
+func newShardLRU(size int) *shardLRU {
+	return &shardLRU{
+		idToNode:   map[enode.ID][]*list.Element{},
+		shardNodes: map[ShardInfo]*list.List{},
+		size:       size,
+		rng:        rand.New(rand.NewSource(rand.Int63())),
+	}
+}
+
+type nodeWithShardInfo struct {
+	key  ShardInfo
+	node *enode.Node
+}
+
+// time complexity: O(number of previous indexes present for node.ID)
+func (l *shardLRU) remove(node *enode.Node) {
+	elements := l.idToNode[node.ID()]
+	for _, element := range elements {
+		key := element.Value.(nodeWithShardInfo).key
+		l.shardNodes[key].Remove(element)
+	}
+	delete(l.idToNode, node.ID())
+}
+
+// if a node is removed for a list, remove it from idToNode too
+func (l *shardLRU) removeFromIdToNode(ele *list.Element) {
+	nodeID := ele.Value.(nodeWithShardInfo).node.ID()
+	for ind, entries := range l.idToNode[nodeID] {
+		if entries == ele {
+			l.idToNode[nodeID] = append(l.idToNode[nodeID][:ind], l.idToNode[nodeID][ind+1:]...)
+			break
+		}
+	}
+	if len(l.idToNode[nodeID]) == 0 {
+		delete(l.idToNode, nodeID)
+	}
+}
+
+func nodeToRelayShard(node *enode.Node) *protocol.RelayShards {
+	shard, err := wenr.RelaySharding(node.Record())
+	if shard == nil || err != nil { // if no shard info, then add to node to Cluster 0, Index 0
```

Comment:
I think this condition is incorrect. We should probably return the error. Looking quickly at the implementation of `RelaySharding` if there's an error, it means that the `enode` contained incorrect information, so I'd probably ignore this peer

Reply:
I think this was added to handle defaultpububTopic?

Reply:
In that case shard would be nil. I'd do something like this, let me know what you think and I'll commit it:

```
func nodeToRelayShard(node *enode.Node) (*protocol.RelayShards, error) {
	shard, err := wenr.RelaySharding(node.Record())
        if err != nil {
           return nil, err
       }

	if shard == nil { // if no shard info, then add to node to Cluster 0, Index 0
		shard = &protocol.RelayShards{
			ClusterID: 0,
			ShardIDs:  []uint16{0},
		}
	}

	return shard, nil
}
```